godot-fmt/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+                    ('#', _) => {
+                        state = FormatState::LastHash;
+                    }
+                    // only used in macros, always prefix
+                    ('$', _) => {
+                        state = FormatState::LastJoinedOperator;
+                    }
+                    ('&', FormatState::LastJoinedOperator) => {
+                        state = FormatState::NothingSpecial;
+                    }
+                    ('&', _) => {
+                        state = FormatState::LastJoinedOperator;
+                    }
+                    ('!', FormatState::LastHash) => {
+                        state = FormatState::LastHash;
+                    }
+                    ('!', FormatState::LastJoinedOperator) => {
+                        state = FormatState::NothingSpecial;
+                    }
+                    ('!', _) => {
+                        state = FormatState::LastJoinedOperator;
+                    }
+                    (':', FormatState::LastJoinedOperator) => {
+                        state = FormatState::LastDoubleColon;
+                    }
+                    (_, FormatState::Start) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    (_, FormatState::NothingSpecial) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    (_, FormatState::LastJoinedOperator) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    (_, FormatState::LastDoubleColon) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    ('.', FormatState::LastIdentifier) => {
+                        state = FormatState::LastJoinedOperator;
+                    }
+                    (_, FormatState::LastIdentifier) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
```

Comment:
All these conditions except for `'.'`  have the same body, could you collapse the match arms with `|`?